subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java

```diff
@@ -213,37 +205,39 @@ public interface ProviderFactory {
     /**
      * Allows lazy access to the output of the external process.
      *
-     * When the process output is read at configuration time it is considered as an input to the
+     * <p>When the process output is read at configuration time it is considered as an input to the
      * configuration model. Consequent builds will re-execute the process to obtain the output and
      * check if the cached model is still up-to-date.
      *
-     * The process input and output streams cannot be configured.
+     * <p>The process input and output streams cannot be configured.
      *
+     * <p>For more sophisticated use cases, like handling execution exceptions, supplying input streams or processing the output in a streaming way, use {@link ValueSource}.
      *
      * @param action the configuration of the external process with the output stream
      * pre-configured.
      * @return an interface that allows lazy access to the process' output.
+     *
      * @since 7.5
      */
-    @Incubating
     ExecOutput exec(Action<? super ExecSpec> action);
 
     /**
      * Allows lazy access to the output of the external java process.
      *
-     * When the process output is read at configuration time it is considered as an input to the
+     * <p>When the process output is read at configuration time it is considered as an input to the
      * configuration model. Consequent builds will re-execute the process to obtain the output and
      * check if the cached model is still up-to-date.
      *
-     * The process input and output streams cannot be configured.
+     * <p>The process input and output streams cannot be configured.
+     *
+     * <p>For more sophisticated use cases, like handling execution exceptions, supplying input streams or processing the output in a streaming way, use {@link ValueSource}.
      *
      * @param action the configuration of the external process with the output stream
      * pre-configured.
      * @return an interface that allows lazy access to the process' output.
      *
      * @since 7.5
      */
-    @Incubating
     ExecOutput javaexec(Action<? super JavaExecSpec> action);
```

Comment:
Seems odd that we have both `ExecOperations` and `ProviderFactory` exposing an `exec` and `javaexec`. 

What are the differences between these two? Can we favor one over the other / potentially deprecate one of them as well?

Reply:
> Seems odd that we have both `ExecOperations` and `ProviderFactory` exposing an `exec` and `javaexec`.
> 
> What are the differences between these two? Can we favor one over the other / potentially deprecate one of them as well?

ExecOperations provides blocking API and can serve as a drop-in replacement for task actions. It is quite convenient and very versatile. Because of this versatility it is impossible to reliably use it at configuration time, because Gradle cannot figure out what is the configuration input and check it for up-to-dateness. ProviderFactory provides special, well, providers, to obtain the process output in a controlled way. It can be used at configuration time, but can also be wired to task inputs as usual.

In theory, we can remove ExecOperations and ask users to use the Java's ProcessBuilder directly, but this lacks some niceties that ExecOperations provide, like output redirections.